### ISSUE TITLE
File UIDs/GIDs in docker image messed up since server-2022-12-01

### ISSUE THREAD

**Issue by rroesch1** [badges/shields#8906]
### Are you experiencing an issue with...

My own instance of shields

### 🐞 Description

The UIDs/GIDs of files in the docker image are messed up since `shieldsio/shields:server-2022-12-01`

```bash
$ docker run -it shieldsio/shields:server-2022-12-01 /bin/sh -c 'ls -lah ./node_modules/*/package.json | egrep -v "root\\s+root"'
-rw-r--r--    1 node     node         551 Dec  1 19:03 ./node_modules/after/package.json
-rw-r--r--    1 501      dialout      413 Dec  1 19:03 ./node_modules/arraybuffer.slice/package.json
-rw-r--r--    1 501      dialout      282 Dec  1 19:03 ./node_modules/backo2/package.json
-rw-r--r--    1 501      dialout     1.0K Dec  1 19:03 ./node_modules/balanced-match/package.json
-rw-r--r--    1 node     node         971 Dec  1 19:03 ./node_modules/base64-arraybuffer/package.json
-rw-r--r--    1 501      dialout      639 Dec  1 19:03 ./node_modules/bintrees/package.json
-rw-r--r--    1 node     node        2.3K Dec  1 19:03 ./node_modules/check-error/package.json
-rw-r--r--    1 501      dialout      381 Dec  1 19:03 ./node_modules/component-bind/package.json
-rw-r--r--    1 502      dialout      356 Dec  1 19:03 ./node_modules/component-inherit/package.json
-rw-r--r--    1 501      dialout      651 Dec  1 19:03 ./node_modules/core-util-is/package.json
-rw-r--r--    1 501      dialout     1.1K Dec  1 19:03 ./node_modules/debug/package.json
-rw-r--r--    1 501      dialout      751 Dec  1 19:03 ./node_modules/decode-uri-component/package.json
-rw-r--r--    1 501      dialout     1.0K Dec  1 19:03 ./node_modules/deep-is/package.json
-rw-r--r--    1 501      dialout     2.7K Dec  1 19:03 ./node_modules/esprima/package.json
-rw-r--r--    1 501      dialout     1.1K Dec  1 19:03 ./node_modules/esutils/package.json
-rw-r--r--    1 501      dialout      998 Dec  1 19:03 ./node_modules/fast-levenshtein/package.json
-rw-r--r--    1 501      dialout      639 Dec  1 19:03 ./node_modules/filter-obj/package.json
-rw-r--r--    1 node     node         670 Dec  1 19:03 ./node_modules/fleau/package.json
-rw-r--r--    1 501      dialout      577 Dec  1 19:03 ./node_modules/fs.realpath/package.json
-rw-r--r--    1 502      dialout     1.5K Dec  1 19:03 ./node_modules/function-bind/package.json
-rw-r--r--    1 node     node        2.4K Dec  1 19:03 ./node_modules/get-func-name/package.json
-rw-r--r--    1 501      dialout      621 Dec  1 19:03 ./node_modules/has-cors/package.json
-rw-r--r--    1 12001236 12001236     744 Dec  1 19:03 ./node_modules/hpack.js/package.json
-rw-r--r--    1 12001236 12001236     710 Dec  1 19:03 ./node_modules/http-deceiver/package.json
-rw-r--r--    1 501      dialout      230 Dec  1 19:03 ./node_modules/indexof/package.json
-rw-r--r--    1 501      dialout      658 Dec  1 19:03 ./node_modules/inflight/package.json
-rw-r--r--    1 501      cdrw         481 Dec  1 19:03 ./node_modules/ip6/package.json
-rw-r--r--    1 501      dialout      786 Dec  1 19:03 ./node_modules/isexe/package.json
-rw-r--r--    1 501      dialout      796 Dec  1 19:03 ./node_modules/json-stringify-safe/package.json
-rw-r--r--    1 node     node         927 Dec  1 19:03 ./node_modules/levn/package.json
-rw-r--r--    1 501      dialout      746 Dec  1 19:03 ./node_modules/lodash.countby/package.json
-rw-r--r--    1 501      dialout      749 Dec  1 19:03 ./node_modules/lodash.defaults/package.json
-rw-r--r--    1 501      dialout      746 Dec  1 19:03 ./node_modules/lodash.groupby/package.json
-rw-r--r--    1 501      dialout      758 Dec  1 19:03 ./node_modules/lodash.isarguments/package.json
-rw-r--r--    1 501      dialout      740 Dec  1 19:03 ./node_modules/lodash.times/package.json
-rw-r--r--    1 501      dialout      901 Dec  1 19:03 ./node_modules/lru_map/package.json
-rw-r--r--    1 1001     1003         444 Dec  1 19:03 ./node_modules/map-values/package.json
-rw-r--r--    1 501      dialout      704 Dec  1 19:03 ./node_modules/ms/package.json
-rw-r--r--    1 node     node         651 Dec  1 19:03 ./node_modules/multilog/package.json
-rw-r--r--    1 501      dialout     1.1K Dec  1 19:03 ./node_modules/node-env-flag/package.json
-rw-r--r--    1 502      dialout      374 Dec  1 19:03 ./node_modules/object-filter/package.json
-rw-r--r--    1 501      dialout      574 Dec  1 19:03 ./node_modules/once/package.json
-rw-r--r--    1 110779   users        914 Dec  1 19:03 ./node_modules/optionator/package.json
-rw-r--r--    1 node     node        1.2K Dec  1 19:03 ./node_modules/prelude-ls/package.json
-rw-r--r--    1 501      dialout      655 Dec  1 19:03 ./node_modules/range_check/package.json
-rw-r--r--    1 node     node        1.0K Dec  1 19:03 ./node_modules/redis-errors/package.json
-rw-r--r--    1 node     node        1.3K Dec  1 19:03 ./node_modules/redis-parser/package.json
-rw-r--r--    1 501      dialout      822 Dec  1 19:03 ./node_modules/select-hose/package.json
-rw-r--r--    1 node     node         659 Dec  1 19:03 ./node_modules/semver-compare/package.json
-rw-r--r--    1 node     node         235 Dec  1 19:03 ./node_modules/socket.io-adapter/package.json
-rw-r--r--    1 501      dialout     1.4K Dec  1 19:03 ./node_modules/spdy/package.json
-rw-r--r--    1 501      dialout      556 Dec  1 19:03 ./node_modules/strict-uri-encode/package.json
-rw-r--r--    1 501      dialout      843 Dec  1 19:03 ./node_modules/tdigest/package.json
-rw-r--r--    1 node     node         795 Dec  1 19:03 ./node_modules/to-array/package.json
-rw-r--r--    1 node     node         873 Dec  1 19:03 ./node_modules/type-check/package.json
-rw-r--r--    1 501      dialout      452 Dec  1 19:03 ./node_modules/ucfirst/package.json
-rw-r--r--    1 501      dialout      694 Dec  1 19:03 ./node_modules/util-deprecate/package.json
-rw-r--r--    1 503      dialout     1.7K Dec  1 19:03 ./node_modules/word-wrap/package.json
-rw-r--r--    1 501      dialout      606 Dec  1 19:03 ./node_modules/wrappy/package.json
-rw-r--r--    1 node     node        1.0K Dec  1 19:03 ./node_modules/xtend/package.json
-rw-r--r--    1 501      dialout      965 Dec  1 19:03 ./node_modules/yeast/package.json
```

Especially the uid 12001236 is a problem since the uid cannot be mapped in a rootless environment (see e.g. moby/moby#43576)

Pulling the image in a rootless environment therefore fails (e.g. rootless docker daemon):
```
…
#5 ERROR: failed to register layer: ApplyLayer exit status 1 stdout:  stderr: failed to Lchown "/usr/src/app/node_modules/hpack.js/.npmignore" for UID 120012366, GID 120012366 (try increasing the number of subordinate IDs in /etc/subuid and /etc/subgid): lchown /usr/src/app/node_modules/hpack.js/.npmignore: invalid argument
------
 > [1/2] FROM docker.io/shieldsio/shields:server-2023-02-01@sha256:7d4df7c168f4ccafcd1b94d64383affc628a45363bb93311bba9c9beb8321693:
```

**Comment by chris48s** [comment]
Hi. I don't think the UIDs/GIDs used in the image are something we explicitly specify. I'm not sure how we would control this tbh. I'm not really familiar with the specifics of this issue/environment so it is going to be hard to advise but I'll offer two suggestions.

1. Reading over the error message thread you've linked it seems like the suggested approach here is to increase the number of subordinate IDs in `/etc/subuid` and `/etc/subgid` on the target system. That may be worth looking into.

2. If the images we push to DockerHub don't work for your use case, we do [tag the GitHub repo](https://github.com/badges/shields/tags) with the same tags we use for Docker Hub images so you could clone the repo, check out your desired tag (e.g: `server-2023-02-01`) and `docker build` your own image to your own specification, although I'm not sure what you would change to get a different result.

**Comment by rroesch1** [comment]
* Regarding the environment: The issue is reproducable in any rootless enviroment.
 For example:
  * Docker in [rootless mode](https://docs.docker.com/engine/security/rootless/)
  * [podman](https://docs.podman.io/en/latest/)
  * Docker nested in [sysbox](https://github.com/nestybox/sysbox)

  e.g. pulling the image with podman fails:
```
$ podman pull docker.io/shieldsio/shields:server-2023-02-01
Trying to pull docker.io/shieldsio/shields:server-2023-02-01...
Getting image source signatures
Copying blob 8921db27df28 done  
Copying blob 2ed4f044afde done  
Copying blob b8243c93ff91 done  
Copying blob 4e08f9bea56d done  
Copying blob 1a58e6c56783 done  
Copying blob f9b23d032312 done  
Copying config f314b42dbf done  
Writing manifest to image destination
Storing signatures
  Error processing tar file(exit status 1): potentially insufficient UIDs or GIDs available in user namespace (requested 120012366:120012366 for /usr/src/app/node_modules/hpack.js/.npmignore): Check /etc/subuid and /etc/subgid: lchown /usr/src/app/node_modules/hpack.js/.npmignore: invalid argument
Error: Error committing the finished image: error adding layer with blob "sha256:1a58e6c56783650ff55e7cba85835074a69dc9d6eac1d0b4bfe729a9386f1173": Error processing tar file(exit status 1): potentially insufficient UIDs or GIDs available in user namespace (requested 120012366:120012366 for /usr/src/app/node_modules/hpack.js/.npmignore): Check /etc/subuid and /etc/subgid: lchown /usr/src/app/node_modules/hpack.js/.npmignore: invalid argument
```

* Increasing the number of subuids is not an option (at least for me). (tbh the default limitation of 2⁶⁴ UIDs should be sufficent)

* Looks the UID/GID is from the tarball served by the npm registry (e.g. https://registry.npmjs.org/hpack.js/-/hpack.js-2.1.6.tgz)
```bash
$ tar -tvf hpack.js-2.1.6.tgz 2>/dev/null 
-rw-r--r-- 120012366/120012366 744 2016-09-18 20:00 package/package.json
-rw-r--r-- 120012366/120012366  28 2016-03-24 17:14 package/.npmignore
-rw-r--r-- 120012366/120012366 1860 2016-03-24 17:14 package/README.md
-rw-r--r-- 120012366/120012366   69 2016-03-24 17:14 package/.travis.yml
-rwxr-xr-x 120012366/120012366  692 2016-03-24 17:14 package/bin/benchmark
-rw-r--r-- 120012366/120012366  397 2016-03-24 17:14 package/lib/hpack.js
-rw-r--r-- 120012366/120012366 3002 2016-09-18 19:59 package/lib/hpack/compressor.js
-rw-r--r-- 120012366/120012366 3857 2016-03-24 17:14 package/lib/hpack/decoder.js
-rw-r--r-- 120012366/120012366 2700 2016-09-18 19:41 package/lib/hpack/decompressor.js
-rw-r--r-- 120012366/120012366 2778 2016-03-24 17:14 package/lib/hpack/encoder.js
-rw-r--r-- 120012366/120012366 12265 2016-03-24 17:14 package/lib/hpack/huffman.js
-rw-r--r-- 120012366/120012366  9806 2016-03-24 17:14 package/lib/hpack/static-table.js
-rw-r--r-- 120012366/120012366  2209 2016-09-18 19:43 package/lib/hpack/table.js
-rw-r--r-- 120012366/120012366   511 2016-03-24 17:14 package/lib/hpack/utils.js
-rw-r--r-- 120012366/120012366  3827 2016-09-18 20:00 package/test/compressor-test.js
-rw-r--r-- 120012366/120012366  5291 2016-03-24 17:14 package/test/decoder-test.js
-rw-r--r-- 120012366/120012366  5248 2016-03-24 17:14 package/test/decompressor-test.js
-rw-r--r-- 120012366/120012366  4541 2016-03-24 17:14 package/test/encoder-test.js
-rw-r--r-- 120012366/120012366  6529 2016-03-24 17:14 package/test/fixtures.js
-rw-r--r-- 120012366/120012366 18148 2016-03-24 17:14 package/tools/gen-huffman.js
-rw-r--r-- 120012366/120012366  1907 2016-03-24 17:14 package/tools/gen-static-table.js
-rw-r--r-- 120012366/120012366   454 2016-03-24 17:14 package/tools/utils.js
```

**Comment by rroesch1** [comment]
@chris48s badges/shields#8908 should fix this
